app/models/spree/order.rb

```diff
@@ -516,6 +516,23 @@ def process_payments!
       errors.add(:base, e.message) && (return result)
     end
 
+    def process_payments_offline!
+      raise Core::GatewayError, Spree.t(:no_pending_payments) if pending_payments.empty?
+
+      pending_payments.each do |payment|
+        break if payment_total >= total
+
+        payment.process_offline!
+
+        if payment.completed?
+          self.payment_total += payment.amount
+        end
+      end
+    rescue Core::GatewayError => e
+      errors.add(:base, e.message)
+      false
+    end
+
```

Comment:
If I'm not mistaken, the only difference with `#process_payments!` is the call to `#process_offline!` instead of `#process!`. In that case, I'd remove the duplication and highlight that slight difference passing the particular strategy as an argument. Something like:

```ruby
def process_payments!(strategy = :online)
(...)
  if strategy == :offline
    payment.process_offline!
  else
    payment.process!
  end
(...)
end
```

Stills, this wouldn't be perfect but I think is a bit easier to extract into a service later on, which I think we should do at some point so we remove endless responsibilities from the order model.

Reply:
I'm aware of the fact that my suggestion :nose: like a boolean argument and that is a code smell. What if we extract the common logic into a method that both `#process_payments!` and `#process_payments_offline!` both use? In the end, we could have consumer code like:

```ruby
def process_payments_offline!
  process_each_payment do |payment|
    payment.process_offline!
  end
end
```

Reply:
Okay, have a look at 4ca354ef848a4239212006530cedeabc69687b92 (included an explanation in the commit description)

Reply:
Nice! that's much better IMO.

Reply:
I agree! Thanks for thinking through it with me.